File: algorithms/2_medium/M/03381_maximum_subarray_sum_with_length_divisible_by_k/solution.py
class Solution:
    def maxSubarraySum(self, nums: List[int], k: int) -> int:
        max_sum = -sys.maxsize - 1
        pref_sum = 0
        min_pref = [sys.maxsize // 2] * k
        min_pref[k - 1] = 0

        for (i, num) in enumerate(nums):
            min_pref_index = i % k

            pref_sum += num
            max_sum = max(max_sum, pref_sum - min_pref[min_pref_index])
            min_pref[min_pref_index] = min(min_pref[min_pref_index], pref_sum)

        return max_sum

File: algorithms/2_medium/M/03381_maximum_subarray_sum_with_length_divisible_by_k/solution.rs
impl Solution {
    pub fn max_subarray_sum(nums: Vec<i32>, k: i32) -> i64 {
        let k_usize = k as usize;
        let mut max = i64::MIN;
        let mut pref_sum = 0_i64;
        let mut min_pref = vec![(i64::MAX) / 2; k_usize];
        min_pref[k_usize - 1] = 0;

        for (i, num) in nums.iter().enumerate() {
            let min_pref_index = i % k_usize;

            pref_sum += (*num) as i64;
            max = max.max(pref_sum - min_pref[min_pref_index]);
            min_pref[min_pref_index] = min_pref[min_pref_index].min(pref_sum);
        }

        max
    }
}
